### PR TITLE
core: handle URL callbacks in coretasks

### LINKED ISSUE #1489: Disabling url.py module also disables @url decorator
The point of adding `@module.url()` to the module API was to simplify the song-and-dance of adding a module's URL patterns to the global list of "callbacks" so that `url.py` wouldn't process links already handled by a more specific plugin.

However, if `url.py` is not loaded, **any** callable using `@url` will simply never run, because these functions are dispatched **in `url.py`** instead of by Sopel's core:

https://github.com/sopel-irc/sopel/blob/0543f2f59ad0ff08b8f22491be6edffcc8694cd4/sopel/modules/url.py#L239-L245

I consider this to be a pretty major bug in the API. The point of having `url_callbacks` is for `url.py` to avoid double-processing links that another module already handles. Module API functionality **must not** depend on loading any module besides `coretasks` (which cannot be disabled, for obvious reasons).

_(On a personal note, this also explains why I couldn't get `@url`-decorated callables to trigger on my own Sopel instance when I tried a few years ago. I've had `url.py` disabled approximately forever.)_

----

I think the solution here is to move `url.py`'s URL-detection and callback-dispatch logic into `coretasks`, and the `exclude` & `exclusion_char` settings from the `[url]` config section into `[core]`. This `@rule`:

https://github.com/sopel-irc/sopel/blob/0543f2f59ad0ff08b8f22491be6edffcc8694cd4/sopel/modules/url.py#L157-L158

would become just another `@url` decorator.

That said, making all modules equal by moving dispatch of `@url` into `coretasks` presents a dilemma: How to determine which callback/pattern is "most appropriate", and avoid having the same URL processed by both `url.py` and (an)other module(s)? I'm almost positive that this conundrum is the reason why the current implementation was implemented as-is, but I'm also sure it's a solvable problem.

Given the lack of immediate ideas (at least on my end), and the fact that this is a fairly major overhaul, I'm assigning this issue to Sopel 7. That's the ideal, but if implementing a solution requires an API change, this will be pushed to Sopel 8 so as to provide advance warning via release notes. (It will also, of course, be pushed off if it simply isn't ready in a reasonable amount of time.)

### PR DESCRIPTION
**Pull request by Exirel**
This fix #1489 - at last!

PR #1508 added a new interface for plugin maintainers, that would help us work with URL callbacks. In this PR, I:

* never trigger URL Callbacks from the `url` built-in plugin
* share the "find URL" feature into `sopel.web` (with unit tests for good measure)
* trigger URL callbacks from a `@url` defined in `coretasks`

This way, one may disable the `url` plugin, and the bot won't auto-title URLs. However, other plugins will still be able to define and use URL Callbacks.

Plugins that use `@rule` and define their own callback will still work, and that won't trigger callbacks.

By the way, this should fix https://github.com/sopel-irc/sopel-github/issues/13 too as soon as the github plugin uses `@url` instead of `@rule`.

### CODE REVIEW

File: sopel/coretasks.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 import time
 import sopel
 import sopel.module
+import sopel.web
```
Comment (dgw) [inline sopel/coretasks.py]: It's unnecessary to import `sopel.submodule` if `sopel` is already imported, unless doing `from sopel.submodule import SpecificThing`.

File: sopel/modules/url.py [hunk 2]
```diff
@@ -226,21 +198,15 @@ def process_urls(bot, trigger, urls):
     return results
 
 
-def check_callbacks(bot, trigger, url, run=True):
+def check_callbacks(bot, trigger, url):
```
Comment (dgw) [inline sopel/modules/url.py]: Update docstring to remove the `run` param too.

File: test/test_web.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    urls = list(search_urls('!http://example.com', exclusion_char='!'))
+    assert not urls, 'Must not find URL, found %d' % len(urls)
+
+    urls = list(
+        search_urls(
+            '!http://example.com http://example.com',
+            exclusion_char='!'
+        )
+    )
+    assert len(urls) == 1, 'Must find 1 URL, found %d' % len(urls)
```
Comment (dgw) [inline test/test_web.py]: This isn't a useful test of the exclusion character, since even if the `exclusion_char` doesn't work properly the `search_urls` function will only return one item anyway (because it only returns unique URLs).

Better to double down on the following test, and assert that `http://a.com` is not in the returned list.

File: sopel/web.py [hunk 4]
```diff
@@ -237,3 +237,39 @@ def iri_to_uri(iri):
     urlencode = urllib.urlencode
 else:
     urlencode = urllib.parse.urlencode
+
+
+def search_urls(text, exclusion_char=None, clean=False):
+    def trim_url(url):
```
Comment (dgw) [inline sopel/web.py]: I think this should either become its own utility function (in `sopel.web` or `sopel.tools`), or become inlined code in the only place where it's used. Defining a function inside another is one of those oddball things that we should probably avoid doing unless there's a good reason.

File: sopel/coretasks.py (hunk 1 above)
Comment (Exirel) [reply sopel/coretasks.py]: I was keeping the style of import used in `coretasks.py`. I don't like it. I'd rather have `from sopel import module, web, [...]`... but I would also keep that for another PR, if you agree.

Comment (dgw) [reply sopel/coretasks.py]: > another PR

Yeah, OK, I'll lay off for now and you can fix a bunch of those all at once. :+1:

File: sopel/web.py (hunk 4 above)
Comment (Exirel) [reply sopel/web.py]: I moved this function mostly "as-is", adding the `exclusion_char` parameter.

Fun fact: since `find_url` was never call with `clean=True`, `trim_url` was (and still is) never called, ever. And I've no idea why.

File: test/test_web.py (hunk 3 above)
Comment (Exirel) [reply test/test_web.py]: I agree that it's a very edge-case, but here is my thought process.

First, this is what we do: ignore **this particular instance** of a URL:

    >>> search_urls('!http://a.com http://b.com http://a.com', '!')
    ['http://b.com', 'http://a.com']

Which is fine by me. But it could mean something else: blacklist **all instance** of a URL:

    >>> search_urls('!http://a.com http://b.com http://a.com', '!')
    ['http://b.com']

I think I should change my test to add this extra `http://b.com` URL, maybe that would speak to you?

Comment (dgw) [reply test/test_web.py]: No, it should _not_ blacklist **all instances** of that URL in the same line, haha.

I think I get it—this test is checking that the `exclusion_char` only applies to the URL it abuts, and not all identical URLs in the line. The previous and next tests check that the `exclusion_char` actually excludes the single URL it abuts.

You needn't change the tests. Just a couple comments about what each is testing (since they're grouped under the same function) would be enough.

File: sopel/web.py (hunk 4 above)
Comment (dgw) [reply sopel/web.py]: > Fun fact: since `find_url` was never call with `clean=True`, `trim_url` was (and still is) never called, ever. And I've no idea why.

I know why. It's because, before the flood of collaborators Sopel gained in the last few months, I left a small but very important change out of #1413 and there was nobody else looking at the code to call me out on my mistake. >.> So that'll be a good (bugfix) bugfix to add to the 6.6.x pile.

For this PR, you'll want to add a similar bugfix in the same place. I haven't written the patch just yet, but it's probably going to involve just passing `clean=True` from `title_auto` in the `url` module, so you can tentatively do the equivalent here.

File: test/test_web.py (hunk 3 above)
Comment (Exirel) [reply test/test_web.py]: Will do! :)

File: sopel/web.py (hunk 4 above)
Comment (dgw) [reply sopel/web.py]: Update: The relevant bugfix is in #1515.

Comment (Exirel) [reply sopel/web.py]: >  so you can tentatively do the equivalent here

I'll add some tests and do this.

Comment (Exirel) [reply sopel/web.py]: I did extract it, I let you see if you agree with that.

**Comment by Exirel** [comment]
@dgw I think I fixed most of the issue, you're welcome for the second round, and once everything's fine I'll rebase/squash accordingly. :)

**Comment by Exirel** [comment]
> multiple spaces before operator

Damn, I used `flake8 sopel` but forgot the test-suite. I was quite in a hurry this morning, my bad!

File: test/test_web.py [hunk 5]
```diff
@@ -93,3 +94,50 @@ def test_search_urls_exclusion_char_only_once():
...
+
+@pytest.mark.parametrize('trailing_char', TRAILING_CHARS)
+@pytest.mark.parametrize('left, right', ENCLOSING_PAIRS)
+def test_trim_url_trailing_char_and_enclosing(trailing_char, left, right):
+    test_url = 'http://exemple.com/'
+    assert test_url == trim_url(test_url + right + trailing_char)
+
+    # assert the trailing char is kept if there is something else
+    test_url = 'http://exemple.com/' + trailing_char
+    assert test_url  == trim_url(test_url + right)
```
Comment (Exirel) [inline test/test_web.py]: The double-space offender is here!

File: test/test_web.py (hunk 3 above)
Comment (Exirel) [reply test/test_web.py]: Done, have a look at the new tests! :)

**Comment by Exirel** [comment]
What the hell is wrong with Python 2.7 this time???

**Comment by dgw** [comment]
Is that really an issue in py2? If so it probably needs to be applied on 6.6.x too :/

**Comment by Exirel** [comment]
> Is that really an issue in py2? If so it probably needs to be applied on 6.6.x too :/

Yes. :sob:

**Comment by dgw** [comment]
> > Is that really an issue in py2? If so it probably needs to be applied on 6.6.x too :/
> 
> Yes. 😭

Well. You'd think I would have caught that myself, given that my primary Sopel instance still runs in py2 and I'm therefore invested in maintaining compatibility…

Added this bugfix to my todo list for 6.6.x. :/

**Comment by Exirel** [comment]
@dgw From my point of view, everything's fine. Yet it appears that Travis isn't happy with an unrelated test (it has something to do with Google stuff).

So if you want to merge, it's ready for the last review.

**Comment by Exirel** [comment]
Py2.7 & Py3.4 tests pass with the exception of the `ip` test, so I think we are good. :)

**Comment by dgw** [comment]
I'll get Travis to re-run the tests after #1523 makes it into `master`. Did you want to do any squashing on this branch at all, @Exirel, or just leave all the commits in? (I'm fine with having the full evolution, just curious if you planned to squash down the little fixups like for flake8 failures and py2.7 compat etc.)

File: sopel/coretasks.py [hunk 6]
```diff
@@ -791,3 +792,20 @@ def track_topic(bot, trigger):
     if channel not in bot.channels:
         return
     bot.channels[channel].topic = trigger.args[-1]
+
+
+@sopel.module.rule(r'(?u).*(https?://\S+).*')
```
Comment (dgw) [inline sopel/coretasks.py]: I wonder about protocol support. Like, there's no reason that plugins shouldn't be able to register handlers for any kind of URL, not just HTTP(S) links, especially since your added `sopel.web.search_urls()` function includes `ftp` in the list of protocols. (Side note: `ftps` exists, too. 😸)

But like, a gaming channel might want a bot to handle `steam://` links, for example.

Could be a separate PR, sure! But I think it's worth putting in a bit of extra work on supporting arbitrary protocols if enabled by the bot owner and/or a plugin—and that means not hard-coding `https?` here.

File: sopel/modules/url.py [hunk 7]
```diff
@@ -92,33 +89,13 @@ def setup(bot):
...
-            urls = [trim_url(url) for url in urls]
-        return urls
-
-    find_urls = find_func
-
-
-@commands('title')
-@example('.title http://google.com', '[ Google ] - google.com')
+@module.commands('title')
+@module.example('.title http://google.com', '[ Google ] - google.com')
```
Comment (dgw) [inline sopel/modules/url.py]: Want to bring this link into the modern era while you're in here?

```suggestion
@module.example('.title https://www.google.com', '[ Google ] - www.google.com')
```

File: sopel/modules/url.py [hunk 8]
```diff
@@ -167,20 +148,22 @@ def title_auto(bot, trigger):
...
         message = '[ %s ] - %s' % (title, domain)
         if tinyurl:
             message += ' ( %s )' % tinyurl
         # Guard against responding to other instances of this bot.
         if message != trigger:
             bot.say(message)
+            bot.memory['last_seen_url'][trigger.sender] = url
+
+        # stop at 4th iteration
+        if i == 3:
```
Comment (dgw) [inline sopel/modules/url.py]: Why is this one (auto-title) `enumerate(process_urls([blah]))` and the other (manual command) is `results[:4]`?

I feel like we shouldn't need an `i` counter variable in Python. :/ But I also think both places should use the same pattern, not have one store all URLs in a `results` list and the other stop iteration before processing all URLs.

File: sopel/web.py [hunk 9]
```diff
@@ -237,3 +237,40 @@ def iri_to_uri(iri):
...
+
+    return url
+
+
+def search_urls(text, exclusion_char=None, clean=False):
+    re_url = r'((?:http|https|ftp)(?::\/\/\S+))'
+    if exclusion_char is not None:
+        re_url = r'((?<!%s)(?:http|https|ftp)(?::\/\/\S+))' % (exclusion_char)
+
+    r = re.compile(re_url, re.IGNORECASE | re.UNICODE)
```
Comment (dgw) [inline sopel/web.py]: It shouldn't be necessary to recompile the regex every time this function is called. Runtime changes to the `exclusion_char` (via `.set`) are the only things that could invalidate it, and at the moment I don't think supporting that is worth the efficiency loss.

Even if it's only running every time a link is found (with the arbitrary protocol support I'd like to add eventually, at least every line containing `://`), it's still not good to recompile regexes all the time for no reason. It's a bad habit. :)

File: sopel/web.py [hunk 10]
```diff
@@ -237,3 +237,40 @@ def iri_to_uri(iri):
...
+
+    urls = re.findall(r, text)
+    if clean:
+        urls = (trim_url(url) for url in urls)
+
+    # yield unique URL in their order of appearance
```
Comment (dgw) [inline sopel/web.py]: ♪ Nitpick time~ ♪

```suggestion
    # yield unique URLs in their order of appearance
```

File: test/test_web.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    assert 'http://b.com' in urls
+
+
+TRAILING_CHARS = list('.,?!\'":;')
+ENCLOSING_PAIRS = [('(', ')'), ('[', ']'), ('{', '}'), ('<', '>')]
+
+
+@pytest.mark.parametrize('trailing_char', TRAILING_CHARS)
+def test_trim_url_remove_trailing_char(trailing_char):
+    test_url = 'http://exemple.com/'
```
Comment (dgw) [inline test/test_web.py]: It doesn't actually _matter_ in the slightest, but above here we have `example.com` and from here to the end of the file it's `exemple.com` instead. 😕

File: sopel/coretasks.py (hunk 6 above)
Comment (Exirel) [reply sopel/coretasks.py]: Hey, I wonder the same thing, because when I decided to kept the old behavior, I found it odd. At this point, I changed so many little things that I think it's safe to change a bit more.

Say I change the rule to `.*`, then perform a search based on a list of protocol provided by configuration. Say `core.url_allowed_schemes`?

File: sopel/modules/url.py (hunk 8 above)
Comment (Exirel) [reply sopel/modules/url.py]: Yeah, but there is a problem:

* we fetch all URLs, no matter how many we display
* we display only 4 of them at most (like it's a common thing)
* in the other function, we try to be smart with an error message if the number of URL does not match the number of result
* one of the function check for malicious links, the other doesn't,
* etc.

I think the differences between these functions are irrelevant and should be restructured into this:

* always display *all* URL, it's not like people will flood with tinyurl on 512 bytes (minus their hostmask)
* display an error message per URL that we fail to get (and not just URL that are handled by a callback)
* apply the same protection on all URLs
* register the "last seen URL" on both .title and auto-url

This way I can fix that `enumerate` problem, and never rely on `list(generator-function)` again. :D

File: sopel/web.py (hunk 9 above)
Comment (Exirel) [reply sopel/web.py]: Don't worry, internaly, `re.compile` use caching: it won't re-process the same pattern twice.

File: test/test_web.py (hunk 11 above)
Comment (Exirel) [reply test/test_web.py]: lol, French being bad at English :D

**Comment by dgw** [comment]
Shall we get back on this horse now that you have a laptop again, @Exirel? :grin:

**Comment by Exirel** [comment]
Oh yes, right. I'll resume working on that very soon.

File: sopel/config/core_section.py [hunk 12]
```diff
@@ -216,3 +216,9 @@ def homedir(self):
 
     verify_ssl = ValidatedAttribute('verify_ssl', bool, default=True)
     """Whether to require a trusted SSL certificate for SSL connections."""
+
+    auto_url_schemes = ListAttribute(
+        'auto_url_schemes',
+        strip=True,
+        default=('http', 'https', 'ftp'))
```
Comment (dgw) [inline sopel/config/core_section.py]: `ListAttribute` should default to a `list`, hmm?
```suggestion
        default=['http', 'https', 'ftp'])
```

File: sopel/config/core_section.py [hunk 13]
```diff
@@ -216,3 +216,9 @@ def homedir(self):
 
     verify_ssl = ValidatedAttribute('verify_ssl', bool, default=True)
     """Whether to require a trusted SSL certificate for SSL connections."""
+
+    auto_url_schemes = ListAttribute(
```
Comment (dgw) [inline sopel/config/core_section.py]: This belongs below the `auth_*` values. The properties defined here are alphabetized.

File: sopel/config/core_section.py [hunk 14]
```diff
@@ -216,3 +216,9 @@ def homedir(self):
...
+        'auto_url_schemes',
+        strip=True,
+        default=('http', 'https', 'ftp'))
+    """List of URL schemes the URL callbacks are triggered."""
```
Comment (dgw) [inline sopel/config/core_section.py]:
```suggestion
    """List of URL schemes that will trigger URL callbacks."""
```

Need to think about whether it's possible, but this should link to documentation for that API feature. Maybe `bot.register_url_callback()`?

File: sopel/web.py [hunk 15]
```diff
@@ -237,3 +237,43 @@ def iri_to_uri(iri):
...
+
+    # clean unmatched parentheses/braces/brackets
+    for (opener, closer) in [('(', ')'), ('[', ']'), ('{', '}'), ('<', '>')]:
+        if url[-1] == closer and url.count(opener) < url.count(closer):
+            url = url[:-1]
+
+    return url
+
+
+def search_urls(text, exclusion_char=None, clean=False, schemes=None):
```
Comment (dgw) [inline sopel/web.py]: I've this feeling that the function name should be `parse_urls` or `find_urls` instead… Thoughts?

Comment (Exirel) [reply sopel/web.py]: I'd argue that `parse_urls` should parse URLs, ie. get a list of URLs and extract something from them. `find_urls` on the other hand would do the same thing as `search_urls`, but I can't see the difference. Would you tell me what your gut feeling are saying?

**Comment by Exirel** [comment]
@dgw I manually tested that everything worked fine. Now I'll try to write some documentation. This can be done in this PR, or in a separated one. As you wish. :)

**Comment by Exirel** [comment]
I added more information in the `sopel.module.url` decorator's docstring, with a link to `auto_url_schemes` and to `bot.register_url_callback`.

Note that I've tested the example added in this docstring to make sure it worked as intented.

File: sopel/config/core_section.py [hunk 16]
```diff
@@ -87,6 +87,18 @@ class CoreSection(StaticSection):
...
+    auto_url_schemes = ListAttribute(
+        'auto_url_schemes',
+        strip=True,
+        default=['http', 'https', 'ftp'])
+    """List of URL schemes that will trigger URL callbacks.
+
+    Used by the URL callbacks feature, see :func:`sopel.module.url` decorator
+    for plugins.
+
+    The default value will allows ``http``, ``https``, and ``ftp``.
```
Comment (dgw) [inline sopel/config/core_section.py]:
```suggestion
    The default value will allow ``http``, ``https``, and ``ftp``.
```

or

```suggestion
    The default value allows ``http``, ``https``, and ``ftp``.
```

File: sopel/config/core_section.py [hunk 17]
```diff
@@ -87,6 +87,18 @@ class CoreSection(StaticSection):
 
     May not apply, depending on ``auth_method``."""
 
+    auto_url_schemes = ListAttribute(
+        'auto_url_schemes',
+        strip=True,
+        default=['http', 'https', 'ftp'])
+    """List of URL schemes that will trigger URL callbacks.
+
+    Used by the URL callbacks feature, see :func:`sopel.module.url` decorator
```
Comment (dgw) [inline sopel/config/core_section.py]:
```suggestion
    Used by the URL callbacks feature; see :func:`sopel.module.url` decorator
```

:smirk:

**Comment by Exirel** [comment]
@dgw ready for a last review to :ship:

**Comment by Exirel** [comment]
Fixup commits squashed, ready to :ship:
